### DISCUSSION TITLE
Middleware Layer

### DISCUSSION THREAD

**Discussion by surgiie** [bashly-framework/bashly#174] [Ideas]
Not sure if this already exists as a feature or if there is a clean way to do this, but it would be nice to define a "middleware" layer on commands that would allow for reusing bash logic for whether the command should run or not, for example, say you have several commands that require docker to  be running, it would be useful if I could just do something like this in a command definition

```yaml
# some command.yml:
name: do_thing
middleware: docker_is_running
...
```

Where `docker_is_running` is just a script in `src/middleware/docker_is_running.sh` or an available function on path called `docker_is_running`

That could look something like:

``` bash
    # Ensure that Docker is running...
    if ! docker info > /dev/null 2>&1; 
    then
        echo "Docker is not running!"
        exit 1
    fi
```
Im basically looking for something similar to arg validation: https://bashly.dannyb.co/advanced/validations/#custom-validations
but at the command execution level if that makes sense. 

This makes it a bit easier to reuse the logic for checking if a command is allowed to be called in your cli command sh files/scripts.

**Comment by DannyBen** [comment]
Well, this functionality does not exist, except the [`command.dependencies`](https://bashly.dannyb.co/configuration/command/#dependencies) configuration which checks that other programs exist.

This can be quite easily achieved by adding a `lib/require_docker.sh` with a function that can be called as the first thing in commands that requires it.

What is the advantage of defining it in the YAML?

```bash
# lib/require_docker.sh
require_docker() {
  # Ensure that Docker is running...
  if ! docker info > /dev/null 2>&1;  then
    echo "Docker is not running!"
    exit 1
  fi
}

# any_command.sh
require_docker
# rest of the command...
```

**Reply by surgiie** [reply]
Ahh perfect that works fine. Missed that in the docs 🤦

**Comment by DannyBen** [comment]
By the way:

1. The library feature I mentioned is in the [Writing Your Scripts > Adding common functions](https://bashly.dannyb.co/usage/writing-your-scripts/#adding-common-functions) section.
2. There is also [`initialize.sh`](https://bashly.dannyb.co/usage/writing-your-scripts/#initialization-function) in case *all* your commands require docker to be running.
3. The `middleware` suggestion is still a good one. Although a middleware is usually also capable of modifying the request, which I think might be too much to require the user to handle, but perhaps a `filter` option might be a possible addition. I am just not sure if it will provide any added value on top of just using `lib` functions.

If you feel we should still explore this route, I am open to it.

**Comment by DannyBen** [comment]
I am testing a `filters` option implementation in #175

**Comment by surgiie** [comment]
Great! Thank you for considering this feature. The lib dependencies works just as fine. I do like the option to define it in the yaml to keep command configuration in a single place but it doesn’t bring any extra value as you mentioned. Just a personal preference. If you decide not to move forward with this, I can still accomplish what I need without it 😄

**Reply by DannyBen** [reply]
Well - I was trigger-happy and implemented it :)

One of bashly's mandates is to prevent your code from being executed if its requirements are not met. It is done with the `dependencies`, `environment_variables` and with custom validations - to name a few "bouncer" features.

So these custom filters fall into the same category.

In the implementation, I added `filters` option to a command, which can accept either an array or a single string.

You can see the example here: https://github.com/DannyBen/bashly/tree/add/filters/examples/filters#filters-example

Unless you have comments about the implementation, I think I will merge it, as it is a helpful feature.

**Reply by surgiie** [reply]
Looks great to me!

**Comment by DannyBen** [comment]
Excellent, thanks for the suggestion - I will release.

**Reply by surgiie** [reply]
Thanks again! 👍 I'll keep an eye out for the release. :)

**Comment by DannyBen** [comment]
This is released as 0.7.4 with documentation [here](https://bashly.dannyb.co/configuration/command/#filters) and example [here](https://github.com/DannyBen/bashly/tree/master/examples/filters#readme).
